src/store.rs

```diff
@@ -453,54 +488,37 @@ impl<F: LurkField> Store<F> {
     pub fn intern_list(&mut self, elts: &[Ptr<F>]) -> Ptr<F> {
         elts.iter()
             .rev()
-            .fold(self.lurk_sym("nil"), |acc, elt| self.intern_cons(*elt, acc))
+            .fold(self.nil_ptr(), |acc, elt| self.intern_cons(*elt, acc))
     }
 
-    pub fn intern_symbol(&mut self, sym: Symbol) -> Ptr<F> {
-        let ptr = if sym.path.is_empty() {
-            Ptr::null(ExprTag::Sym)
-        } else {
-            let mut ptr = self.symnil();
-            for s in sym.path.iter() {
-                let str_ptr = self.intern_string(s);
-                ptr = self.intern_symcons(str_ptr, ptr);
-            }
-            if sym == Symbol::nil() {
-                Ptr {
-                    tag: ExprTag::Nil,
-                    raw: ptr.raw,
-                    _f: ptr._f,
-                }
-            } else if sym.is_keyword() {
-                Ptr {
-                    tag: ExprTag::Key,
-                    raw: ptr.raw,
-                    _f: ptr._f,
-                }
-            } else {
-                ptr
-            }
-        };
-        self.symbol_cache.insert(sym, Box::new(ptr));
-        ptr
+    pub fn intern_symbol_path(&mut self, path: &[String]) -> Ptr<F> {
+        path.iter().fold(self.symnil(), |acc, s| {
+            let s_ptr = self.intern_string(s);
+            self.intern_symcons(s_ptr, acc)
+        })
     }
 
-    pub fn get_sym(&self, sym: &Symbol) -> Option<Ptr<F>> {
-        let ptr = self.symbol_cache.get(sym).cloned()?;
-        if *sym == Symbol::nil() {
-            Some(Ptr {
-                tag: ExprTag::Nil,
-                raw: ptr.raw,
-                _f: ptr._f,
-            })
-        } else {
-            Some(*ptr)
+    pub fn intern_symbol(&mut self, sym: &Symbol) -> Ptr<F> {
+        match self.symbol_cache.get(sym) {
+            Some(ptr) => *ptr,
+            None => {
+                use crate::tag::ExprTag::{Key, Nil};
+                let path_ptr = self.intern_symbol_path(sym.path());
+                let sym_ptr = if sym == &lurk_sym("nil") {
+                    path_ptr.cast(Nil)
+                } else if !sym.is_keyword() {
+                    path_ptr
+                } else {
+                    path_ptr.cast(Key)
+                };
```

Comment:
The negation in the condition makes this harder to parse than strictly necessary, at least for me.